src/OpenTelemetry.Contrib.PersistentStorage/LocalFileStorage.cs

```diff
@@ -0,0 +1,177 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Timers;
+
+namespace OpenTelemetry.Contrib.PersistentStorage
+{
+    /// <summary>
+    /// Persistent file storage <see cref="LocalFileStorage"/> allows to save data
+    /// as blobs in file storage.
+    /// </summary>
+    public class LocalFileStorage : IPersistentStorage, IDisposable
+    {
+        private readonly string directoryPath;
+        private readonly long maxSizeInBytes;
+        private readonly long retentionPeriodInMilliseconds;
+        private readonly int writeTimeoutInMilliseconds;
+        private readonly Timer maintenanceTimer;
+        private bool disposedValue;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="LocalFileStorage"/>
+        /// class.
+        /// </summary>
+        /// <param name="path">
+        /// Sets file storage folder location where blobs are stored.
+        /// </param>
+        /// <param name="maxSizeInBytes">
+        /// Maximum allowed storage folder size.
+        /// Default is 50 MB.
+        /// </param>
+        /// <param name="maintenancePeriodInMilliseconds">
+        /// Maintenance event runs at specified interval.
+        /// Removes expired leases and blobs that exceed retention period.
+        /// Default is 2 minute.
+        /// </param>
+        /// <param name="retentionPeriodInMilliseconds">
+        /// Retention period in milliseconds for the blob.
+        /// Default is 2 days.
+        /// </param>
+        /// <param name="writeTimeoutInMilliseconds">
+        /// Controls the timeout when writing a buffer to blob.
+        /// Default is 1 minute.
+        /// </param>
+        public LocalFileStorage(
+                                string path,
+                                long maxSizeInBytes = 52428800,
+                                int maintenancePeriodInMilliseconds = 6000,
+                                long retentionPeriodInMilliseconds = 172800,
+                                int writeTimeoutInMilliseconds = 6000)
+        {
+            if (path == null)
+            {
+                throw new ArgumentNullException(nameof(path));
+            }
+
+            this.directoryPath = PersistentStorageHelper.CreateSubdirectory(path);
```

Comment:
We need to validate the input range (e.g. if `maintenancePeriodInMilliseconds` is a crazy number like every 1ms or negative).